Cpu strided complex unaryop

dylanbespalko opened the pull request:
@ezyang 
In-tree changes to pytorch to support complex numbers are being submitted here.
Out-of-tree support for complex numbers is here: [pytorch-cpu-strided-complex extension](https://gitlab.com/pytorch-complex/pytorch-cpu-strided-complex)

Changes in this PR.

- [x]  Added complex support of UnaryOp kernels.

I have fixed several problems that prevented the complex number kernels from re-using existing code:
1. Added fake AVX support that allows complex kernels to call `cpu_kernel_vec` instead of `cpu_kernel`.
  - `aten/src/ATen/cpu/vec256/vec256_complex_double.h`
  - `aten/src/ATen/cpu/vec256/vec256_complex_double.h`
  - Code was taken from @Roger-luo complex number repo.
  - I also added `real`, `imag` methods to all other data types and `vec256_base.h`
2. Added `abs`, `real`, `imag` utility functions that operate on complex dtypes and act as no-ops for other data types. 
  - 'aten/src/ATen/native/cpu/ZMath.h` contains these utility functions.

One final problem exists: `std::complex` does not support several compare operators. Would it be acceptable to replace occurrences of `a > b` with `gt(a, b)`, where `gt` is implemented similar to `real_impl` and `imag_impl` in `ZMath.h`?

dylanbespalko committed:
Added registration and type promotion for ComplexFloat and ComplexDouble Types

dylanbespalko committed:
Added registration and type promotion for ComplexFloat and ComplexDouble Types

dylanbespalko committed:
Updated python_arg_parser to support complex args and added complex dispatch switch statements

dylanbespalko committed:
Updated python_arg_parser to support complex args and added complex dispatch switch statements

dylanbespalko committed:
Added support for complex scalar ops

dylanbespalko committed:
Added support for complex scalar ops

dylanbespalko committed:
Added support for copy and binary ops

dylanbespalko committed:
Added support for copy and binary ops

dylanbespalko committed:
Finalized support for copy and binary ops

dylanbespalko committed:
Merge branch 'cpu-strided-complex-binaryop' of github.com:dylanbespalko/pytorch into cpu-strided-complex-binaryop

dylanbespalko committed:
Merge branch 'master' of github.com:pytorch/pytorch into cpu-strided-complex-unaryop

dylanbespalko committed:
Added complex number support for most unary ops

dylanbespalko committed:
merged changes from pytorch master

dylanbespalko committed:
debugging AVX related issues on aarm64

dylanbespalko committed:
debugging AVX related issues on aarm64

dylanbespalko committed:
Added fixes for aarm64

dylanbespalko committed:
Merge branch 'master' of github.com:pytorch/pytorch into cpu-strided-complex-unaryop

dylanbespalko committed:
Finalizing complex number support for UnaryOps

ezyang commented:
> One final problem exists: std::complex does not support several compare operators. Would it be acceptable to replace occurrences of a > b with gt(a, b), where gt is implemented similar to real_impl and imag_impl in ZMath.h?

I have a broader question to ask: how exactly are you defining comparison on complex numbers? This is one of those cases where I feel we should NOT provide an implementation.

ezyang commented:
@gchanan if you could look at the dispatch macro bits

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -131,6 +131,40 @@
+  variants: function, method
+  named_guard: False
+
+- func: real_(Tensor(a!) self) -> Tensor(a!)
```
`real_` changes the dtype of a tensor, so I'm not sure how you can actually do it inplace...

dylanbespalko replied:
These should absolutely NOT be inplace. Thanks for catching this.

ezyang commented on aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:
```diff
@@ -82,12 +101,21 @@ static void bitwise_not_kernel(TensorIterator& iter) {
+          iter,
+          [=](scalar_t a) -> scalar_t {
+            return a - scalar_t(std::trunc(std::real(a)), std::trunc(std::imag(a))); });
+    });
```
lol, I'm not sure if this is ever actually a useful operation on complex numbers, but sure

dylanbespalko replied:
I agree, I will remove it. I was using it early on because it was an easy function that doesn't all `cpu_vec`.

ezyang commented on aten/src/ATen/cpu/vec256/vec256_complex_double.h:
```diff
@@ -0,0 +1,328 @@
+  Vec256(std::complex<double> val) {
+    for (int i = 0; i != size(); i++) {
+      values[i] = val;
+    }
```
These operations haven't been vectorized at all! The loops aren't even unrolled!

In other words, what it seems like what you've implemented here is not the equivalent of `vec256_float.h` (which is all vectorized) but `vec256_base.h` (which provides non-vectorized implementations of all operations to default back to.) So, I have two questions:

1. Can we rename the header so this is clear?
2. Can we deduplicate `vec256_complex_{double,float}`, since if these are non-vectorized, the implementations should basically be the same. This is similar to how `base` is only written once.

ezyang reviewed (changes requested):
This looks pretty good. However, inplace real/imag, and non-vectorized header, are the show stoppers for landing.

dylanbespalko commented:
> @gchanan if you could look at the dispatch macro bits

@gchanan: I'm trying NOT to support half precision complex numbers right now, but I found that `eps`, `min`, `max`, and `tiny` require that I implement `AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND_HALF`. This is the only place in the code where this dispatch macro is needed.

dylanbespalko committed:
Applied the recomendations of PR #25850

dylanbespalko committed:
Merge branch 'master' of github.com:pytorch/pytorch into cpu-strided-complex-unaryop

dylanbespalko committed:
Fixing CI failure due to DL_RUNTIME_BUG and incomplete docs

dylanbespalko commented:
@ezyang 
 
I have renamed the unvectorized code to `vec256_complex_double_no_vec.h` and `vec256_complex_float_no_vec.h`. I will likely combine complex float and complex double after I have a chance to implement some AVX vectorizations.

I have removed the inplace calls to `real_` and `imag_`.

**I have to run CI again to fix a workaround for a bug in `Glibc2.23` called `DL_RUNTIME_BUG` that didn't work for complex numbers.**

Finally, I have made changes based on the recommendations that you made above. I plan to add comparison ops next, however I will only add the operators that are supported by std::complex until I have a better sense of how kernels like `torch.clamp` become vectorized.

I will contact you when the CI is passing.

dylanbespalko commented:
> I have a broader question to ask: how exactly are you defining comparison on complex numbers? This is one of those cases where I feel we should NOT provide an implementation.

I would define a `a > b` as `abs(a) > abs(b)` where `abs() = sqrt(a.real()**2 + a.imag()**2)`.  I work with complex numbers every day and I have never seen another interpretation in real life, however I have noticed that numpy defines `a > b` as `a.real() > b.real()`. Maybe there is ambiguity???

In a library like Pytorch I could see this issue propagating when people define compounded operations like `torch.clamp`. Anyways, I won't implement it until I have a better understanding of how an operation like `torch.clamp` is vectorized for complex numbers.

dylanbespalko committed:
Fixing CI failures related to DL_RUNTIME_BUG

ezyang commented:
> I would define a a > b as abs(a) > abs(b) where abs() = sqrt(a.real()**2 + a.imag()**2). I work with complex numbers every day and I have never seen another interpretation in real life, however I have noticed that numpy defines a > b as a.real() > b.real()

This is kind of what I have expected, but Wikipedia says that this doesn't follow some laws (which I also believe). It is surprising that Numpy only compares reals haha.

dylanbespalko committed:
fixing CI failures related to DL_RUNTIME_BUG

dylanbespalko committed:
still trying to fix the CI failure related to DL_RUNTIME_BUG

dylanbespalko committed:
still debugging CI problem

dylanbespalko committed:
still debugging CI problem

dylanbespalko committed:
still debugging CI problem

dylanbespalko commented:
Still trying to figure out what went wrong with the CI. I'm going to have to step though the changes tomorrow.

vadimkantorov commented:
This should fix https://github.com/pytorch/pytorch/issues/22764. Also sometimes abs squared is needed (which can save sqrt + square ops and bring more numerical stability). `abs(squared = True)` would be nice, but maybe not so generic. In general, Lp norm to p'th power is not unusual. 

Is abs on complex type faster than norm(dim = -1) on a float type? If yes, maybe abs should be called underneath in norm if last dimension of a float tensor is 2.

dylanbespalko commented:
@vadimkantorov 

Yes, I agree. It seems the reason why some complex operators are not implemented in std::complex is that you will find other ways to optimize the computation. Note: this change does not support vectorization yet, but I will work on that as soon as I figure out this CI problem.

dylanbespalko commented on aten/src/ATen/Dispatch.h:
```diff
@@ -139,6 +139,21 @@ inline void deprecated_AT_DISPATCH_ALL_TYPES_AND_HALF_AND_COMPLEX() {}
+    }                                                                                                            \
+  }()
+
+ #define AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND_HALF(TYPE, NAME, ...)                                        \
```
This sounds fine. I'll make the change.

dylanbespalko commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -239,9 +239,9 @@ inline void propagate_names_if_namedtensor_enabled(Tensor& result, const Tensor&
-    checkBackend(#op, result, Backend::device);                        \
     auto iter = TensorIterator::unary_op(result, self,                 \
       /*check_mem_overlap=*/true);                                     \
+    AT_ASSERT(iter.device_type() == DeviceType::CPU);                  \
```
Ok, I think I get it.
`checkDeviceType()` is only checking device.
`checkBackend()` is checking both device and layout.

gchanan commented:
@VitalyFedyunin can you comment on the vec256 changes?

dylanbespalko commented:
@VitalyFedyunin, @gchanan,

I wouldn't spend too much time looking at my Vec256 changes. @ezyang has given me some help on that and I should have a Vec256 complex number implementation soon.

VitalyFedyunin commented:
> @VitalyFedyunin, @gchanan,
> 
> I wouldn't spend too much time looking at my Vec256 changes. @ezyang has given me some help on that and I should have a Vec256 complex number implementation soon.

Got it, add me when you feel ready to receive Vec256 feedback. But unless you actually start using avx2 instructions it is kind of over complicated to add two more classes when you can just make cpu_kernel_vec to ignore vectorization in case of complex types.
